### PR TITLE
Error messages: use a simple heuristic to retrieve sourcefile name from compiled interface

### PR DESCRIPTION
**Pull request by Octachron**
Whenever we compare the module type of a cmi file against the module type of its implementation, the name of the cmi file briefly appears in the error message, for instance

> File "common/a.ml", line 1:        
Error: The implementation common/a.ml
       does not match the interface common/.l.objs/byte/l__A.cmi: 
       The type `t' is required but not provided
       File "common/a.mli", line 1, characters 0-6: Expected declaration

An important observation at this point is that this cmi file name is soon replaced by the interface source file name `common/a.mli` as soon as we are not at the toplevel of the error message.

That seems to be a clear sign that we should avoid speaking temporarily of the cmi file at all at the toplevel ... if we can avoid it.

Part of the issue is that the interface source file name is not that straightforward to determine:

- `cmi` files only records the name of the compiled module
- `dune` compiles `ml` file with `-intf-suffix .ml`, thus we cannot deduce the interface file name from the name of the ml file at all.
- line directives can change the source file name to avoid exposing generated files (e.g. `#<n> "parser.mly") in the error message.

As a compromise this PR proposes to look at the locations of all toplevel signature items from the cmi file. If all those locations share a common filename part, we can sensibly use that filename as the name of the source file for the interface. Otherwise, for chimeric cmi assembled from multiple source files, it seems fine to keep using the cmi filename.

Another less hackish alternative would be to add a flag for `ocamlc` and `ocamlopt` to give a source file for interface files and record that name in the cmi file. That might be a better long term path, but that would require more cooperation with build tools. At the same time, having a more explicit flag might be a better option than using `-intf-suffix .ml` for them. 
(cc @rgrinberg ?)

### CODE REVIEW

**Comment by gasche** [comment]
I think that getting a filename from the locations in the interface is a reasonable approach.

On the other hand, I don't understand the dune-related part of the your explanation:
- Can you explain what is the purpose of `-intf-suffix .ml` and how that affects the error message displayed in your example?
- If I call `ocamlc -c test.mli test.ml` with `val x : bool` and `let x = 42` I also observe `test.cmi` mentioned in the error message, is your heuristic really related to dune?

**Comment by Octachron** [comment]
Sorry, I kept too much of the context in my own mind:
In brief, when compiling an implementation file, the typechecker checks if the interface source file exists by looking for a file named `<name without extension> ^ intf_suffix`. 

One may think that we could reuse this name in the error message itself. However, dune uses `-intf-suffix .ml` to bypass this check when compiling `ml` files. As a side-effect this means that we cannot hope to compute the source file of the interface from the implementation source file.

The heuristic implemented in this PR thus choose to bypass the problem entirely by reading the contents of the cmi file, which also has the advantage of handling line directive in a consistent way.

**Comment by gasche** [comment]
Since 5.0, we have a `-cmi-file foo.cmi` option that lets a build system compiling a `.ml` file specify which interface it should be checked against. Ignoring backward-compatibility concerns for now, would that be a suitable replacement to Dune's `-intf-suffix .ml` hack? If yes, and if Dune was to magically use this, would we still follow the approach in your PR or could we use something simpler/different?

**Comment by Octachron** [comment]
With `-cmi-file`, I expect that the `-intf-suffix .ml` trick is no longer necessary, which would open a path to a version of this PR that would be simpler but also slightly more brittle due to possible inconsistencies between the locations recorded in the cmi and the inferred name of the mli file.

Overall, I think that this PR implementation is the more consistent option and it seems reasonable in term of implementation complexity.

In other words, in this alternative `-cmi-file` word, this PR may not have been my first iteration but we may have ended with this PR at the end.

**Comment by gasche** [comment]
Thanks, this agrees with my intuition. Then I would suggest rephrasing the explanations in your PR to not make it sound like dune is the problem, but instead that our heuristics to finding .mli file are too fragile.

Your description of "chimeric .cmi" makes them sound like a very unlikely situation; I wonder if we can get those in a "natural" way thanks to ppx metaprograms. For example, does ppx_import result in "chimeric" cmi files, or are we safe because ppx_import includes definitions from other files but the toplevel items remain attribute to the importing file?

**Comment by Octachron** [comment]
In fact, there are no ppx modules needed for chimeric cmi: I had forgotten that items imported through an included module type keep their original locations. Thus
```ocaml
(* a.mli *)
include B_intf.S
include C_intf.S
```
is enough to create a chimeric cmi file from the point of view of this PR.

More strangely, if one exclusively include module types from the same mli file
```
include B_intf.S_one
include B_intf.S_two
```
this PR ends up considering that the source file for `B_intf` is the true source file of the interface whih may lead to peculiar error messages (but which are more self-consistent than current error message):
```ocaml
Error: The implementation a.ml does not match the interface b.ml: 
       The module `B1' is required but not provided
       File "b.ml", line 3, characters 4-21: Expected declaration
       The type `t' is required but not provided
       File "b.ml", line 4, characters 4-10: Expected declaration
```

**Comment by gasche** [comment]
I thought of the following alternatives:

1. We could stop trying to give a path to the interface and simply say `The implementation foo.ml is not consistent with its interface` in the short description of the error.
2. We could store in the .cmi an additional `cmi_path` component that is a user-provided (see below) file path stored at cmi-production time

How to infer a file path at cmi-production time? My suggestion would be to use the filename of the Lexing position at the *end* of mli (or ml) parsing. My reasoning is that preprocessing *inclusions* (the ones that result in chimeric locations) typically use lexer directives in a well-bracketed way, where say

```
(* in orig.mli *)
#include foo.mli
```

gets rewritten into

```
(* orig.mli *)
#2 foo.mli
...
#3 origin.mli
```

therefore it makes sense to say that the "last filename" (at the name of the token stream) "wins" and claims to be the user-provided (in some sense) filename.

**Comment by gasche** [comment]
(My current thinking is that the approach (2) that I propose above is better than the current proposal, and could possibly be generated to other build artifacts.)

**Comment by Octachron** [comment]
I agree that your option `2` is probably a better alternative and that updating the cmi file to have a "source path" metadata is probably the best option when taking in account that `include` easily import items from other source files when looking at the types. 

Since I don't expect the new implementation to share much with the current one, let me close this PR.

**Comment by shindere** [comment]
Including source paths in binary artifacts indeed sounds like a nice approach, but doesn't this go against reproducible builds? Or would we then introduce a stripping stage to get rid of this information between compilation and installation?
